src/commands/keygen.rs

```diff
@@ -0,0 +1,341 @@
+use domain::base::zonefile_fmt::ZonefileFmt;
+use domain::sign::{common, GenerateParams};
+use domain::validate::Key;
+use lexopt::Arg;
+
+use crate::error::{Context, Error};
+use crate::parse::parse_name;
+
+use super::{parse_os, parse_os_with, Command, LdnsCommand};
+
+#[derive(Clone, Debug, Args)]
+pub struct Keygen {
+    /// The signature algorithm to generate for
+    #[arg(short = 'a', value_name = "ALGORITHM", value_enum)]
+    algorithm: AlgorithmArg,
+
+    /// Generate a key signing key instead of a zone signing key
+    #[arg(short = 'k')]
+    make_ksk: bool,
+
+    /// The length of the key (for RSA keys only)
+    #[arg(short = 'b', value_name = "BITS", default_value_t = 2048)]
+    bits: u32,
+
+    /// The randomness source to use for generation
+    #[arg(short = 'r', value_name = "DEVICE", default_value_t = String::from("/dev/urandom"))]
+    random: String,
+
+    /// Create symlinks '.key' and '.private' to the generated keys
+    #[arg(short = 's')]
+    #[cfg(target_family = "unix")]
+    create_symlinks: bool,
+
+    /// Overwrite existing symlinks (for use with '-s')
+    #[arg(short = 'f')]
+    #[cfg(target_family = "unix")]
+    force_symlinks: bool,
+
+    /// The domain name to generate a key for
+    #[arg(value_name = "domain name", value_parser = ValueParser::new(parse_name))]
+    name: Name<Vec<u8>>,
+}
+
+const LDNS_HELP: &str = "\
+ldns-keygen -a <algorithm> [-b bits] [-r /dev/random] [-s] [-f] [-v] domain
+  generate a new key pair for domain
+  -a <alg>	use the specified algorithm (-a list to show a list)
+  -k		set the flags to 257; key signing key
+  -b <bits>	specify the keylength
+  -r <random>	specify a random device (defaults to /dev/random)
+		to seed the random generator with
+  -s		create additional symlinks with constant names
+  -f		force override of existing symlinks
+  -v		show the version and exit
+  The following files will be created:
+    K<name>+<alg>+<id>.key	Public key in RR format
+    K<name>+<alg>+<id>.private	Private key in key format
+    K<name>+<alg>+<id>.ds	DS in RR format (only for DNSSEC KSK keys)
+  The base name (K<name>+<alg>+<id> will be printed to stdout\
+";
+
+impl LdnsCommand for Keygen {
+    const HELP: &'static str = LDNS_HELP;
+
+    fn parse_ldns<I: IntoIterator<Item = OsString>>(args: I) -> Result<Self, Error> {
+        let mut algorithm = None;
+        let mut make_ksk = false;
+        let mut bits = 2048;
+        let mut random = String::from("/dev/urandom");
+        let mut create_symlinks = false;
+        let mut force_symlinks = false;
+        let mut name = None;
+
+        let mut parser = lexopt::Parser::from_args(args);
+
+        while let Some(arg) = parser.next()? {
+            match arg {
+                Arg::Short('a') => {
+                    algorithm = Some(parse_os_with("algorithm (-a)", &parser.value()?, |s| {
+                        AlgorithmArg::from_str(s, true)
+                    })?);
+                }
```

Comment:
ldns:
```
$ ldns-keygen -a 8 -a 13
The -a argument can only be used once
```

Reply:
Ooh, thanks!  I'll check this for other kinds of options too.